aya/src/programs/links.rs

```diff
@@ -148,11 +146,49 @@ impl Link for FdLink {
     }
 
     fn detach(self) -> Result<(), ProgramError> {
-        unsafe { close(self.fd) };
+        // detach is a noop since it consumes self. once self is consumed,
+        // drop will be triggered and the link will be detached.
         Ok(())
     }
 }
 
+impl Drop for FdLink {
+    fn drop(&mut self) {
+        if let Some(fd) = self.fd.take() {
+            // Safety: libc
+            unsafe { close(fd) };
+        }
+    }
+}
+
+/// A pinned file descriptor link.
+///
+/// This link has been pinned to the BPF filesystem. On drop, the file descriptor that backs
+/// this link will be closed. Whether or not the program remains attached is dependent
+/// on the presence of the file in BPFFS.
+#[derive(Debug)]
+pub struct PinnedLink {
+    inner: FdLink,
+    path: Option<PathBuf>,
+}
+
+impl PinnedLink {
+    fn new(path: PathBuf, fd: RawFd) -> Self {
+        PinnedLink {
+            inner: FdLink::new(fd),
+            path: Some(path),
+        }
+    }
+
+    /// Removes the pinned link from the filesystem and returns an [`FdLink`].
+    pub fn unpin(mut self) -> Result<FdLink, io::Error> {
+        if let Some(path) = self.path.take() {
```

Comment:
unpin consumes self so you don't need the option? it can only be called once which means the path will always be there